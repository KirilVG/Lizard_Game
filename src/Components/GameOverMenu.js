import React from "react";
import { useState } from "react";
import * as myConstants from "./Constants";
import 'bootstrap/dist/css/bootstrap.min.css';
import { createTheme } from '@mui/material/styles';
import { ThemeProvider } from "@mui/material";
import { Stack } from "react-bootstrap";

function GameOverMenu(props) {
  const [gameScoreSaved, setGameScoreSaved] = useState(false);

  const clickContinue = () => {
    props.onClick();
  };

  const saveScore = () => {
    let lanesAsString = String(props.end.lanesNum);

    let scoreboard = JSON.parse(localStorage.getItem("scoreboard"));
    if (!scoreboard) scoreboard = {};

    let currentScores = scoreboard[lanesAsString];
    if (!currentScores) currentScores = [];

    currentScores.push({
      username: props.end.username,
      lanesNum: props.end.lanesNum,
      score: props.end.score,
      causeOfDeath: props.end.message,
      discMode: myConstants.useDiscoMode,
    });

    currentScores.sort((a, b) => b.score - a.score);

    if (currentScores.length > myConstants.maximumNumberOfScores)
      currentScores.pop();

    scoreboard[lanesAsString] = currentScores;

    localStorage.setItem("scoreboard", JSON.stringify(scoreboard));

    setGameScoreSaved(true);
  };

  const displayScores = () => {
    let lanesAsString = String(props.end.lanesNum);
    let scoreboard = JSON.parse(localStorage.getItem("scoreboard"));

    if (!scoreboard[lanesAsString]) return <li>no current scores</li>;

    let arr = [];

    for (let i = 0; i < scoreboard[lanesAsString].length; i++) {
      arr.push(
        <li key={i} >
            <div style={{backgroundColor: "white", borderRadius:"1vh"}}>
                <h1 style={{color: "black",fontSize:"2vh"}}>{`${scoreboard[lanesAsString][i].username}`}</h1>
                <p style={{color: "black",fontSize:"1.5vh", padding:"0.5vh" }}>
                    {`score: ${scoreboard[lanesAsString][i].score}, cause of death:${scoreboard[lanesAsString][i].causeOfDeath}`}
                </p>
            </div>
        </li>
      );
    }
    return arr;
  };
  const theme = createTheme({
    palette: {
      primary: {
        main: '#ffffff',
      },
      secondary: {
        main: '#000000',
      },
    },
  });

  return (
    <div>
      <div
        className="bg-immage d-flex justify-content-center align-items-center"
        style={{
          backgroundImage: `url("${myConstants.gameOverMenuBackgroundIMGUrl}")`,
          backgroundSize: "cover",
          backgroundRepeat: "no-repeat-y",
          height: `100vh`,
        }}
      >
        <div class="mask" style={{ backgroundColor: "rgba(0, 0, 0, 0.6)" }}>
          <div class="d-flex justify-content-center align-items-center h-100">
            <div class="container">
              <div class="row justify-content-center">
                <div class="col-12 col-md-10 col-lg-7 col-xl-6">
                  <div
                    class="card"
                    style={{
                      backdropFilter: "blur(15px)",
                      backgroundColor: "rgba(255,255,255,.2)",
                      borderRadius: "3em",
                      border: "2px solid rgba(255,255,255,.1)",
                      backgroundClip: "padding-box",
                      boxShadow: "10px 10px 10px rgba(46, 54, 68, 0.03)",
                    }}
                  >
                    <div class="card-body p-3 text-white">
                      <div class="my-4">
                        <h2 class="text-center mb-5">
                          Game Over!
                        </h2>
                        <form>
                          <ThemeProvider theme={theme}>
                            <Stack spacing={3} style={{ alignItems: "center" }} >
                                <div style={{paddingBottom:"3vh"}}>
                                <label>{`Username:${props.end.username}, cause of death:${props.end.message}, score: ${props.end.score}`}</label>
                                </div>

                              {!gameScoreSaved ? (
                                <div style={{paddingBottom:"3vh"}}>
                                <button
                                type="button"
                                class="btn btn-light"
                                style={{
                                  height: `7vh`,
                                  width: "20vh",
                                  fontSize: "2vh",
                                }}
                                onClick={saveScore}
                              >
                                save Score
                              </button>
                              </div>
                              ) : null}
                              <button
                                type="button"
                                class="btn btn-light"
                                style={{
                                  height: `7vh`,
                                  width: "20vh",
                                  fontSize: "2vh",
                                }}
                                onClick={clickContinue}
                              >
                                Play Again
                              </button>
                            </Stack>
                          </ThemeProvider>
                        </form>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="col-12 col-md-10 col-lg-7 col-xl-6">
                  <div
                    class="card"
                    style={{
                      backdropFilter: "blur(15px)",
                      backgroundColor: "rgba(255,255,255,.2)",
                      borderRadius: "3em",
                      border: "2px solid rgba(255,255,255,.1)",
                      backgroundClip: "padding-box",
                      boxShadow: "10px 10px 10px rgba(46, 54, 68, 0.03)",
                    }}
                  >
                    <div class="card-body p-1 text-white">
                      <div class="my-4">
                        <h2 class="text-center mb-5">Scores</h2>
                        <ThemeProvider theme={theme}>
                          <Stack spacing={5} style={{ display: "flex", alignItems: "center" }} >

                            <ol>{displayScores()}</ol>

                          </Stack>
                        </ThemeProvider>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default GameOverMenu;
/*
<div>
      <h1>Game over</h1>
      <label>{`Username:${props.end.username}, cause of death:${props.end.message}, score: ${props.end.score}`}</label>
      <div>
        {!gameScoreSaved ? (
          <button onClick={saveScore}>Save Score</button>
        ) : null}
        <button onClick={clickContinue}>PLay Again</button>
      </div>
      <div>
        <ol>{displayScores()}</ol>
      </div>
    </div>*/
