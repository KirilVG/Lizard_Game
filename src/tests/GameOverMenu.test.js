import { fireEvent, render, screen } from "@testing-library/react";
import GameOverMenu from "../Components/GameOverMenu";
import renderer from "react-test-renderer";

const mockedGameOverInf = {
  score: 100,
  cause: "killed by sth",
  username: "Person",
  lanesNum: 3,
};

const mockedOnClick = () => {};

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = String(value);
  }

  removeItem(key) {
    delete this.store[key];
  }
}

it("game end renders correctly", () => {
  global.localStorage = new LocalStorageMock();

  const tree = renderer
    .create(<GameOverMenu onClick={mockedOnClick} end={mockedGameOverInf} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("game end should be able to save score", () => {
  global.localStorage = new LocalStorageMock();
  const expectedValue = 1;
  render(<GameOverMenu onClick={mockedOnClick} end={mockedGameOverInf} />);
  const buttonEl = screen.getByRole("button", {
    name: /save Score/i,
  });

  fireEvent.click(buttonEl);

  expect(
    JSON.parse(global.localStorage.getItem("scoreboard"))["3"].length
  ).toBe(expectedValue);
});

it("game end should save only 10 scores per game mode", () => {
  global.localStorage = new LocalStorageMock();
  const scoreBoard = {};
  const currentScores = [];
  const expectedValue = 10;
  render(<GameOverMenu onClick={mockedOnClick} end={mockedGameOverInf} />);
  const buttonEl = screen.getByRole("button", {
    name: /save Score/i,
  });
  for (let i = 0; i < 10; i++) {
    currentScores.push({
      username: "joko",
      lanesNum: 3,
      score: 500,
      causeOfDeath: "IDK",
      discMode: false,
    });
  }
  scoreBoard["3"] = currentScores;
  global.localStorage.setItem("scoreboard", JSON.stringify(scoreBoard));

  fireEvent.click(buttonEl);

  expect(
    JSON.parse(global.localStorage.getItem("scoreboard"))["3"].length
  ).toBe(expectedValue);
});

it("game end should execute the given parameter function when play again is pressed", () => {
  global.localStorage = new LocalStorageMock();
  let val = false;
  const expectedValue = true;
  const givenF = () => {
    val = true;
  };
  render(<GameOverMenu onClick={givenF} end={mockedGameOverInf} />);
  const buttonEl = screen.getByRole("button", {
    name: /play again/i,
  });

  fireEvent.click(buttonEl);

  expect(val).toBe(expectedValue);
});
