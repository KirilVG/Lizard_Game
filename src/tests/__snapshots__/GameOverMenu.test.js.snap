// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game end renders correctly 1`] = `
<div
  className="bg-image d-flex justify-content-center align-items-center backgroundIMG"
  style={
    Object {
      "backgroundImage": "url(\\"https://images.unsplash.com/photo-1573595015625-ac674ed1d155?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1000&q=80\\")",
    }
  }
>
  <div
    className="mask darkMask"
  >
    <div
      className="justify-content-center align-items-center h-100 container"
    >
      <div
        className="row d-flex justify-content-center cardHolder"
      >
        <div
          className="col-12 col-md-10 col-lg-7 col-xl-6"
        >
          <div
            className="card glassCard card-body p-3 text-white my-4"
          >
            <h2
              className="text-center mb-5"
            >
              Game Over!
            </h2>
            <div
              className="formStack vstack minBreakpoint-xs"
              spacing={3}
            >
              <div>
                <label>
                  Username:Person, cause of death:undefined, score: 100
                </label>
              </div>
              <div>
                <button
                  className="btn btn-light playButton"
                  onClick={[Function]}
                  type="button"
                >
                  save Score
                </button>
              </div>
              <div>
                <button
                  className="btn btn-light playButton"
                  onClick={[Function]}
                  type="button"
                >
                  Play Again
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col-12 col-md-10 col-lg-7 col-xl-6"
        >
          <div
            className="card glassCard card-body p-1 text-white my-4"
          >
            <h2
              className="text-center mb-5"
            >
              Scores
            </h2>
            <div
              className="formStack vstack minBreakpoint-xs"
              spacing={5}
            >
              <ol
                className="scoreArea"
              >
                <div>
                  <h1>
                    No current scores!
                  </h1>
                  <p>
                    be the first
                  </p>
                </div>
              </ol>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
